### PR TITLE
(LK-C-6) Add controlled generator (multi-qubit) support for Lightning Kokkos

### PR DESCRIPTION
**Pull request by josephleekl**
### Before submitting

Please complete the following checklist when submitting a PR:

- [ ] All new features must include a unit test.
      If you've fixed a bug or added code that should be tested, add a test to the
      [`tests`](../tests) directory!

- [ ] All new functions and code must be clearly commented and documented.
      If you do make documentation changes, make sure that the docs build and
      render correctly by running `make docs`.

- [ ] Ensure that the test suite passes, by running `make test`.

- [ ] Add a new entry to the `.github/CHANGELOG.md` file, summarizing the
      change, and including a link back to the PR.

- [ ] Ensure that code is properly formatted by running `make format`. 

When all the above are checked, delete everything above the dashed
line and fill in the pull request template.

------------------------------------------------------------------------------------------------------------

**Context:**
This PR adds support for 2/4-qubit controlled generator (e.g. IsingXX / SingleExcitation / DoubleExcitation) and MultiRZ to Lightning Kokkos.
**Description of the Change:**

**Benefits:**

**Possible Drawbacks:**

**Related GitHub Issues:**

[sc-76778]

### CODE REVIEW

File: pennylane_lightning/core/src/simulators/lightning_kokkos/gates/BasicGeneratorFunctors.hpp [hunk 1]
```diff
@@ -530,6 +530,386 @@ void applyNCGenGlobalPhase(
...
+        PL_ASSERT(num_qubits >= nw_tot);
+
+        std::vector<std::size_t> all_wires;
+        all_wires.reserve(nw_tot);
+        all_wires.insert(all_wires.begin(), wires.begin(), wires.end());
+        all_wires.insert(all_wires.begin(), controlled_wires.begin(),
+                         controlled_wires.end());
+
+        std::tie(parity, rev_wires) =
+            reverseWires(num_qubits, wires, controlled_wires);
```
Comment (AmintorDusko) [inline pennylane_lightning/core/src/simulators/lightning_kokkos/gates/BasicGeneratorFunctors.hpp:783]: This is a good candidate for a structured binding.
```suggestion
        auto [parity, rev_wires] = reverseWires(num_qubits, wires, controlled_wires);
```

Comment (josephleekl) [reply pennylane_lightning/core/src/simulators/lightning_kokkos/gates/BasicGeneratorFunctors.hpp]: changed to use structured binding now, thanks!

**Review by AmintorDusko — COMMENTED** [review]
So far, so good.

File: pennylane_lightning/core/src/simulators/lightning_kokkos/gates/BasicGeneratorFunctors.hpp [hunk 2]
```diff
@@ -342,7 +339,7 @@ template <class ExecutionSpace, class PrecisionT>
 void applyGenMultiRZ(Kokkos::View<Kokkos::complex<PrecisionT> *> arr_,
                      std::size_t num_qubits,
                      const std::vector<std::size_t> &wires,
-                     [[maybe_unused]] bool inverse = false) {
+                     [[maybe_unused]] const bool inverse = false) {
```
[the remark this answers is not in the document]
Comment (josephleekl) [reply pennylane_lightning/core/src/simulators/lightning_kokkos/gates/BasicGeneratorFunctors.hpp:342]: Good catch, I meant to remove the const from all of them. I'll remove these!

File: pennylane_lightning/core/src/simulators/lightning_kokkos/gates/BasicGeneratorFunctors.hpp [hunk 3]
```diff
@@ -525,6 +522,394 @@ void applyNCGenGlobalPhase(
...
+    const std::vector<std::size_t> &controlled_wires,
+    const std::vector<bool> &controlled_values,
+    const std::vector<std::size_t> &wires,
+    [[maybe_unused]] const bool inverse = false) {
+    auto core_function = KOKKOS_LAMBDA(
+        Kokkos::View<Kokkos::complex<PrecisionT> *> arr, std::size_t i00,
+        std::size_t i01, std::size_t i10, std::size_t i11) {
+        arr(i00) = 0.0;
+        arr(i01) *= Kokkos::complex<PrecisionT>{0.0, 1.0};
+        arr(i10) *= Kokkos::complex<PrecisionT>{0.0, -1.0};
```
[the remark this answers is not in the document]
Comment (josephleekl) [reply pennylane_lightning/core/src/simulators/lightning_kokkos/gates/BasicGeneratorFunctors.hpp:688]: The reason it's done explicitly there is that we're swapping the real and imaginary part of two variables, so we have to copy it and then swap. But here no swapping is required, so we don't need new variables. Or do you mean doing this
```
arr(i01) = Kokkos::complex<PrecisionT>{0.0, imag(i01)};
arr(i10) = Kokkos::complex<PrecisionT>{0.0, -imag(i10)};
```
? Will this be beneficial? :thinking:

**Review by maliasadi — APPROVED** [review]
LGTM!

File: pennylane_lightning/core/src/simulators/lightning_kokkos/gates/BasicGeneratorFunctors.hpp (hunk 3 above)
[the remark this answers is not in the document]
Comment (AmintorDusko) [reply pennylane_lightning/core/src/simulators/lightning_kokkos/gates/BasicGeneratorFunctors.hpp:688]: Yes, this was what I meant. I choose my example poorly.

File: pennylane_lightning/core/src/simulators/lightning_kokkos/gates/BasicGeneratorFunctors.hpp (hunk 2 above)
[the remark this answers is not in the document]
Comment (AmintorDusko) [reply pennylane_lightning/core/src/simulators/lightning_kokkos/gates/BasicGeneratorFunctors.hpp]: Thanks!

File: pennylane_lightning/core/src/simulators/lightning_kokkos/gates/BasicGeneratorFunctors.hpp (hunk 3 above)
[the remark this answers is not in the document]
Comment (AmintorDusko) [reply pennylane_lightning/core/src/simulators/lightning_kokkos/gates/BasicGeneratorFunctors.hpp:688]: Thank you! Feel free to resolve the conversation when you are done.

**Review by AmintorDusko — APPROVED** [review]
Nice job! Thanks!

**Pull request merged**